proposals/text/5390-embedders.md

```diff
@@ -0,0 +1,273 @@
+
+    def embed(self, data: str, ... inference params ... ) -> np.ndarray:
+        # compute embedding
+        return embedding
+
+
+class OpenAIEmbeddingService:
+    ... same as above ...
+```
+
+Given that EmbeddingServices are created through a factory method, when you request an instance, if another identical exists, the method returns that instance instead of a new one.
+
+This makes model reusability automatic in all cases, which can save lots of memory without asking the user to think about it.
+
+- A `(Text/Table/Image/Audio)Embedder` component that does nothing but “wrapping” an EmbeddingService.
+**Part of the public API**.
+```python
+@component
+class HFTextEmbedder:
+
+    class Input:
+        data: List[str]
+
+    class Output
+        embeddings: List[np.ndarray]
+
+    def __init__(self, model_name: str, ... init params ...):
+        self.model_name = model_name
+        self.model_params = ... params ...
+
+    def warm_up(self):
+        self.embedding_service = HFEmbeddingService(self.model_name, **self.model_params)
+
+    def run(self, data):
+        return self.output(self.embedding_service.embed(data.data))
+```
+
+- A DocumentEmbedder component that handles the documents and the offloads the computation to an embedder.
+**Part of the public API**.
+
+```python
+@component
+class HFDocumentEmbedder(HFTextEmbedder):
+  """
+  Note: in this toy example inheritance from HFTextEmbedder makes sense because
+  init and warm_up are identical. If they would differ significantly, let's remove
+  the inheritance to simplify the architecture: it's not mandatory for the rest
+  of the system to work.
+  """
+
+    class Input:
+        documents: List[Document]
+
+    class Output
+        documents: List[np.ndarray]
+
+    def run(self, data):
+        text_strings = [document.content for document in data.documents]
+        embeddings = self.embedding_service.embed(text_strings)
+        documents_with_embeddings = [Document.from_dict(**doc.to_dict, "embedding": emb) for doc, emb in zip(documents, embeddings)]
+        return self.output(documents = documents_with_embeddings)
+```
+
+# Drawbacks
+
+The drawbacks of concept separation between Retrievers and Embedders were discussed in [this proposal](https://github.com/deepset-ai/haystack/blob/main/proposals/text/4370-documentstores-and-retrievers.md) and mainly consist of **migration effort**.
+
+Several ideas were discussed (mostly concerning how to handle queries and Documents in the Embedders) and this seems to be the best approach. However, there are some potential cons to consider:
+  - Proliferation of classes (though they will be small and easy to maintain).
+  - Users need to properly understand which models are appropriate for a task or for embedding queries vs. Documents (see [Different models in the same embedding/retrieval task](#different-models-in-the-same-embeddingretrieval-task)). On the other hand, this can help raise their awareness.
```

Comment:
```suggestion
  - Users need to know which models are appropriate for which task (e.g. embedding queries rather than embedding documents, see [Different models in the same embedding/retrieval task](#different-models-in-the-same-embeddingretrieval-task)). On the other hand, this approach is more explicit and will help making users aware of problems and tradeoffs related to the topic.
```